File: src/pci/virtio/rng/rng.c
#include "rng.h"
#include "mm/mmu/mmu.h"
#include "mmio/mmio.h"
#include "uart/uart.h"
#include "utils/utils.h"

/* Page-aligned backing memory for the virtqueue */
static struct virtq_desc rng_desc[VIRTQ_MAX_SIZE]
    __attribute__((aligned(4096)));
static struct virtq_avail rng_avail __attribute__((aligned(4096)));
static struct virtq_used rng_used __attribute__((aligned(4096)));

/* Buffer for the random byte */
static uint8_t rand_byte;

struct virtio_rng rng_dev;

static void request_byte() {
  /* Submit a request for 1 random byte.
   * The descriptor addr must be a physical address for DMA. */
  uart_println("[RNG] Requesting random byte...");
  uint64_t rand_byte_pa = VIRT_TO_PHYS((uint64_t)(uintptr_t)&rand_byte);
  virtqueue_submit(&rng_dev.vq, rand_byte_pa, 1, VIRTQ_DESC_F_WRITE);
  virtqueue_notify(&rng_dev.vq);

  uart_println("[RNG] Waiting for device...");
  virtqueue_poll(&rng_dev.vq);

  uart_println("[RNG] Got response!");
  uart_printf("Random byte: %x\n", (uint32_t)rand_byte);
}

void pci_virtio_rng_init() {
  uart_println("[RNG] Initializing Device");

  /* Step 0: Find device */
  if (!pci_find_device(VIRTIO_RNG_VENDOR_ID, VIRTIO_RNG_DEVICE_ID,
                       &rng_dev.pci)) {
    uart_errorln("[RNG] Device not found");
    return;
  }

  uart_println("[RNG] Device found");

  // Header Type register[6:0] = header layout type
  // https://wiki.osdev.org/PCI#Header_Type_Register
  if ((pci_get_header_type(&rng_dev.pci) & 0x7F) != PCI_ENDPOINT_DEV_TYPE) {
    uart_errorln("[RNG]: Unexpected header type");
    return;
  }

  pci_assign_bars(&rng_dev.pci);
  pci_enable_device(&rng_dev.pci);
  virtio_parse_capabilities(&rng_dev.pci, &rng_dev.pci_caps);

  /* VirtIO Device Init Sequence
   * All register accesses go through the MMIO layer. */
  uintptr_t base = rng_dev.pci_caps.common_cfg;

  /* Step 1: Reset Device */
  uart_println("[RNG][VIRTIO-INIT][1] Reset Device");
  mmio_write8(base + VIRTIO_COMMON_STATUS, VIRTIO_STATUS_RESET);
  dsb_sy();

  /* Wait for RESET to complete */
  while (mmio_read8(base + VIRTIO_COMMON_STATUS) != VIRTIO_STATUS_RESET) {
  }
  uart_println("[RNG][VIRTIO-INIT][1] Reset Device Complete");

  /* Step 2: ACK */
  uart_println("[RNG][VIRTIO-INIT][2] Ack");
  uint8_t status = mmio_read8(base + VIRTIO_COMMON_STATUS);
  mmio_write8(base + VIRTIO_COMMON_STATUS, status | VIRTIO_STATUS_ACKNOWLEDGE);
  dsb_sy();

  /* Step 3: Set Driver status */
  uart_println("[RNG][VIRTIO-INIT][3] Driver Status");
  status = mmio_read8(base + VIRTIO_COMMON_STATUS);
  mmio_write8(base + VIRTIO_COMMON_STATUS, status | VIRTIO_STATUS_DRIVER);
  dsb_sy();

  /* Step 4: Feature Negotiation */
  uart_println("[RNG][VIRTIO-INIT][4] Negotiate Features");

  /* Read device features - low 32 bits */
  mmio_write32(base + VIRTIO_COMMON_DFSELECT, 0);
  dsb_sy();
  uint32_t feat_lo = mmio_read32(base + VIRTIO_COMMON_DF);
  uart_printf(" Device features[0]: %x\n", feat_lo);

  /* Read device features - high 32 bits */
  mmio_write32(base + VIRTIO_COMMON_DFSELECT, 1);
  dsb_sy();
  uint32_t feat_hi = mmio_read32(base + VIRTIO_COMMON_DF);
  uart_printf(" Device features[1]: %x\n", feat_hi);

  /*
   * The RNG device has no device-specific features.
   * accept VIRTIO_F_VERSION_1 (bit 32, in feat_hi bit 0).
   */
  uint32_t guest_lo = 0;
  uint32_t guest_hi = feat_hi & 0x01;

  mmio_write32(base + VIRTIO_COMMON_GFSELECT, 0);
  dsb_sy();
  mmio_write32(base + VIRTIO_COMMON_GF, guest_lo);
  dsb_sy();
  mmio_write32(base + VIRTIO_COMMON_GFSELECT, 1);
  dsb_sy();
  mmio_write32(base + VIRTIO_COMMON_GF, guest_hi);
  dsb_sy();

  uart_printf(" Accepted Features: lo=%x hi=%x\n", guest_lo, guest_hi);

  /* Step 5: FEATURES_OK */
  status = mmio_read8(base + VIRTIO_COMMON_STATUS);
  mmio_write8(base + VIRTIO_COMMON_STATUS, status | VIRTIO_STATUS_FEATURES_OK);
  dsb_sy();

  /* Step 6: Re-read and verify FEATURES_OK stuck */
  status = mmio_read8(base + VIRTIO_COMMON_STATUS);
  if (!(status & VIRTIO_STATUS_FEATURES_OK)) {
    uart_errorln("[RNG] FEATURES_OK failed");
    return;
  }
  uart_printf("[RNG] Status: %x\n", (uint32_t)status);
  uart_println("[RNG] FEATURES_OK !");

  /* Step 6: Setup virtqueue 0 */
  rng_dev.vq.desc = rng_desc;
  rng_dev.vq.avail = &rng_avail;
  rng_dev.vq.used = &rng_used;

  if (virtqueue_setup(base, 0, &rng_dev.vq, &rng_dev.pci_caps) != ESUCCESS) {
    uart_errorln("[RNG] Virtqueue setup failed");
    return;
  }

  /* Step 7: DRIVER_OK */
  status = mmio_read8(base + VIRTIO_COMMON_STATUS);
  mmio_write8(base + VIRTIO_COMMON_STATUS, status | VIRTIO_STATUS_DRIVER_OK);
  dsb_sy();
  uart_println("[RNG] DRIVER_OK set");

  request_byte();

  return;
}

File: src/virtio/rng.rs
//! VirtIO RNG driver — full device init + random byte generation.
//!
//! Port of `src/pci/virtio/rng/rng.c`.

use super::virtqueue::{self, Virtqueue, VirtqAvail, VirtqDesc, VirtqUsed, VIRTQ_DESC_F_WRITE, VIRTQ_MAX_SIZE};
use super::*;
use crate::mm::mmu::virt_to_phys;
use crate::mmio;
use crate::pci;
use crate::sync::Racy;
use crate::uart;

pub const VIRTIO_RNG_VENDOR_ID: u16 = 0x1AF4;
pub const VIRTIO_RNG_DEVICE_ID: u16 = 0x1044;

const RNG_BUF_SIZE: usize = 256;

/// Page-aligned DMA backing memory for the virtqueue + bounce buffer.
#[repr(C, align(4096))]
struct RngBacking {
    desc: [VirtqDesc; VIRTQ_MAX_SIZE],
    avail: VirtqAvail,
    used: VirtqUsed,
    buf: [u8; RNG_BUF_SIZE],
}

static BACKING: Racy<RngBacking> = Racy::new(RngBacking {
    desc: [VirtqDesc {
        addr: 0,
        len: 0,
        flags: 0,
        next: 0,
    }; VIRTQ_MAX_SIZE],
    avail: VirtqAvail {
        flags: 0,
        idx: 0,
        ring: [0; VIRTQ_MAX_SIZE],
    },
    used: VirtqUsed {
        flags: 0,
        idx: 0,
        ring: [virtqueue::VirtqUsedElem { id: 0, len: 0 }; VIRTQ_MAX_SIZE],
    },
    buf: [0; RNG_BUF_SIZE],
});

struct RngDev {
    caps: VirtioPciCaps,
    vq: Virtqueue,
    ready: bool,
}

static RNG: Racy<RngDev> = Racy::new(RngDev {
    caps: VirtioPciCaps {
        common_cfg: 0,
        notify_base: 0,
        isr_cfg: 0,
        device_cfg: 0,
        notify_off_multiplier: 0,
    },
    vq: Virtqueue::empty(),
    ready: false,
});

pub fn init() {
    uart::println("[RNG] Initializing Device");
    let mut dev = match pci::find_device(VIRTIO_RNG_VENDOR_ID, VIRTIO_RNG_DEVICE_ID) {
        Some(d) => d,
        None => {
            uart::errorln("[RNG] Device not found");
            return;
        }
    };
    if pci::header_type(&dev) & 0x7F != pci::PCI_ENDPOINT_DEV_TYPE {
        uart::errorln("[RNG] Unexpected header type");
        return;
    }
    pci::assign_bars(&mut dev);
    pci::enable_device(&dev);

    let rng = unsafe { RNG.get() };
    pci::config_read16(dev.bus as u16, dev.slot, dev.func, pci::PCI_STATUS); // touch
    super::parse_capabilities(&dev, &mut rng.caps);

    let base = rng.caps.common_cfg;

    // VirtIO init sequence.
    mmio::write8(base + VIRTIO_COMMON_STATUS, VIRTIO_STATUS_RESET);
    dsb_sy();
    while mmio::read8(base + VIRTIO_COMMON_STATUS) != VIRTIO_STATUS_RESET {}

    let mut status = mmio::read8(base + VIRTIO_COMMON_STATUS);
    mmio::write8(base + VIRTIO_COMMON_STATUS, status | VIRTIO_STATUS_ACKNOWLEDGE);
    dsb_sy();
    status = mmio::read8(base + VIRTIO_COMMON_STATUS);
    mmio::write8(base + VIRTIO_COMMON_STATUS, status | VIRTIO_STATUS_DRIVER);
    dsb_sy();

    // Feature negotiation: accept only VIRTIO_F_VERSION_1 (bit 32).
    mmio::write32(base + VIRTIO_COMMON_DFSELECT, 1);
    dsb_sy();
    let feat_hi = mmio::read32(base + VIRTIO_COMMON_DF);
    let guest_hi = feat_hi & 0x01;
    mmio::write32(base + VIRTIO_COMMON_GFSELECT, 0);
    dsb_sy();
    mmio::write32(base + VIRTIO_COMMON_GF, 0);
    dsb_sy();
    mmio::write32(base + VIRTIO_COMMON_GFSELECT, 1);
    dsb_sy();
    mmio::write32(base + VIRTIO_COMMON_GF, guest_hi);
    dsb_sy();

    status = mmio::read8(base + VIRTIO_COMMON_STATUS);
    mmio::write8(base + VIRTIO_COMMON_STATUS, status | VIRTIO_STATUS_FEATURES_OK);
    dsb_sy();
    status = mmio::read8(base + VIRTIO_COMMON_STATUS);
    if status & VIRTIO_STATUS_FEATURES_OK == 0 {
        uart::errorln("[RNG] FEATURES_OK failed");
        return;
    }

    // Point the virtqueue at the static backing memory.
    let backing = unsafe { BACKING.get() };
    rng.vq.desc = backing.desc.as_mut_ptr();
    rng.vq.avail = &mut backing.avail as *mut VirtqAvail;
    rng.vq.used = &mut backing.used as *mut VirtqUsed;

    if !virtqueue::setup(base, 0, &mut rng.vq, &rng.caps) {
        uart::errorln("[RNG] Virtqueue setup failed");
        return;
    }

    status = mmio::read8(base + VIRTIO_COMMON_STATUS);
    mmio::write8(base + VIRTIO_COMMON_STATUS, status | VIRTIO_STATUS_DRIVER_OK);
    dsb_sy();
    rng.ready = true;
    uart::println("[RNG] DRIVER_OK set");
}

/// Fill `dst` with random bytes from the device. Returns bytes produced.
pub fn read(dst: &mut [u8]) -> usize {
    let rng = unsafe { RNG.get() };
    if !rng.ready {
        return 0;
    }
    let backing = unsafe { BACKING.get() };
    let buf_pa = virt_to_phys(backing.buf.as_ptr() as u64);
    let mut done = 0usize;
    while done < dst.len() {
        let chunk = core::cmp::min(dst.len() - done, RNG_BUF_SIZE);
        virtqueue::submit(&mut rng.vq, buf_pa, chunk as u32, VIRTQ_DESC_F_WRITE);
        virtqueue::notify(&rng.vq);
        virtqueue::poll(&mut rng.vq);
        dst[done..done + chunk].copy_from_slice(&backing.buf[..chunk]);
        done += chunk;
    }
    done
}
